ndk-context/src/lib.rs

```diff
@@ -0,0 +1,86 @@
+//! but could also be initialized by java or kotlin code when embedding in an existing android
+//! project.
+//!
+//! ```no_run
+//! let ctx = ndk_context::android_context();
+//! let vm = unsafe { jni::JavaVM::from_raw(ctx.vm() as *mut jni::sys::JavaVM) }?;
+//! let env = vm.attach_current_thread();
+//! let class_ctx = env.find_class("android/content/Context")?;
+//! let audio_service = env.get_static_field(class_ctx, "AUDIO_SERVICE", "Ljava/lang/String;")?;
+//! let audio_manager = env
+//!     .call_method(
+//!         ctx.context() as jni::sys::jobject,
+//!         "getSystemService",
+//!         "(Ljava/lang/String;)Ljava/lang/Object;",
+//!         &[audio_service],
+//!     )?
+//!     .l()?;
+//! ```
+use std::ffi::c_void;
+
+static mut ANDROID_CONTEXT: Option<AndroidContext> = None;
+
+/// [`AndroidContext`] provides the pointers required to interface with the jni on android
+/// platforms.
+#[derive(Clone, Copy, Debug)]
+pub struct AndroidContext {
+    java_vm: *mut c_void,
+    context_jobject: *mut c_void,
+}
+
+impl AndroidContext {
+    /// A handle to the `JavaVM` object.
+    ///
+    /// Usage with [__jni__](https://crates.io/crates/jni) crate:
+    /// ```no_run
+    /// let ctx = ndk_context::android_context();
+    /// let vm = unsafe { jni::JavaVM::from_raw(ctx.vm() as *mut jni::sys::JavaVM) }?;
+    /// let env = vm.attach_current_thread();
+    /// ```
+    pub fn vm(self) -> *mut c_void {
+        self.java_vm
+    }
+
+    /// A handle to an [android.content.Context](https://developer.android.com/reference/android/content/Context).
+    /// In most cases this will be a ptr to an `Activity`, but this isn't guaranteed.
+    ///
+    /// Usage with [__jni__](https://crates.io/crates/jni) crate:
+    /// ```no_run
+    /// let ctx = ndk_context::android_context();
+    /// let vm = unsafe { jni::JavaVM::from_raw(ctx.vm() as *mut jni::sys::JavaVM) }?;
+    /// let env = vm.attach_current_thread();
+    /// let class_ctx = env.find_class("android/content/Context")?;
+    /// let audio_service = env.get_static_field(class_ctx, "AUDIO_SERVICE", "Ljava/lang/String;")?;
+    /// let audio_manager = env
+    ///     .call_method(
+    ///         ctx.context() as jni::sys::jobject,
+    ///         "getSystemService",
+    ///         "(Ljava/lang/String;)Ljava/lang/Object;",
+    ///         &[audio_service],
+    ///     )?
+    ///     .l()?;
+    /// ```
+    pub fn context(self) -> *mut c_void {
+        self.context_jobject
+    }
+}
+
+/// Main entry point to this crate. Returns an [`AndroidContext`].
+pub fn android_context() -> AndroidContext {
+    unsafe { ANDROID_CONTEXT.expect("android context was initialized") }
+}
+
+/// Initializes the [`AndroidContext`]. [`AndroidContext`] is initialized by `ndk-glue` before
+/// main is called.
+///
+/// # Safety
+///
+/// The pointers need to be valid.
+pub unsafe fn initialize_android_context(java_vm: *mut c_void, context_jobject: *mut c_void) {
```

Comment:
well, it's unsafe to call after main was called (as it can spawn threads and cause data races etc) also it's unsafe to not be initialized before main. I think `#[doc(hidden)]` as originally proposed is probably the right thing to do.